nl/nl_linux.go

```diff
@@ -622,16 +622,17 @@ func (s *NetlinkSocket) Receive() ([]syscall.NetlinkMessage, error) {
 	if fd < 0 {
 		return nil, fmt.Errorf("Receive called on a closed socket")
 	}
-	rb := make([]byte, RECEIVE_BUFFER_SIZE)
-	nr, _, err := unix.Recvfrom(fd, rb, 0)
+	var rb [RECEIVE_BUFFER_SIZE]byte
+	nr, _, err := unix.Recvfrom(fd, rb[:], 0)
 	if err != nil {
 		return nil, err
 	}
 	if nr < unix.NLMSG_HDRLEN {
 		return nil, fmt.Errorf("Got short response from netlink")
 	}
-	rb = rb[:nr]
-	return syscall.ParseNetlinkMessage(rb)
+	rb2 := make([]byte, nr)
+	copy(rb2, rb[:nr])
+	return syscall.ParseNetlinkMessage(rb2)
```

Comment:
why not simply pass the slice of the stack version at this point?

Reply:
If you return `rb[:nr]` then `rb` "escapes" the stack and must be allocated on the heap (http://www.agardner.me/golang/garbage/collection/gc/escape/analysis/2015/10/18/go-escape-analysis.html) - i.e. what makes `rb` allocated on the stack is not only the use of array instead of slice, but also not returning any reference to it.
Therefore if you try to return `rb` instead of `rb2` the performance will decline back where it was as 64K must be allocated on the heap instead of the minimum necessary number of bytes need to contain the response. Allocation on the stack and copying are much cheaper operations than memory management on the heap in general.